notecard/main.go

```diff
@@ -1,107 +1,109 @@
 package main
 
 import (
-	"encoding/json"
 	"fmt"
+	"io"
+	"log"
 	"net/http"
 	"os"
 
 	"github.com/blues/note-go/notecard"
 )
 
-type Transport string
-
 const (
-	Serial Transport = "serial"
-	I2C    Transport = "i2c"
+	Serial = "serial"
+	I2C    = "i2c"
 )
 
-var card *notecard.Context
+type server struct {
+	card      *notecard.Context
+	initError error
+}
 
-func serveNotecard(w http.ResponseWriter, req *http.Request) {
+var transport = os.Getenv("NOTECARD_TRANSPORT")
 
-	if req.Method != http.MethodPost {
-		w.WriteHeader(http.StatusMethodNotAllowed)
-		fmt.Fprintf(w, "Method not allowed")
-		return
-	}
-
-	var data map[string]interface{}
-	if err := json.NewDecoder(req.Body).Decode(&data); err != nil {
-		http.Error(w, fmt.Sprintf("Error decoding JSON: %v", err), http.StatusBadRequest)
-		return
-	}
+func handleError(w http.ResponseWriter, err error, msg string) {
+	err_str := fmt.Sprintf("%s: %v", msg, err)
+	http.Error(w, err_str, http.StatusInternalServerError)
+	log.Print(err_str)
+}
 
-	for key, value := range data {
-		fmt.Printf("%s: %v\n", key, value)
+func (s *server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if s.card == nil || s.initError != nil {
+		handleError(w, s.initError, "while initialising notecard")
+		log.Fatal("Notecard not initialised, exiting...")
 	}
```

Comment:
There's two stances here: either you trust that the caller have consistently and properly set the (s.card, s.initError) tuple, or you don't.

If you do, it means you only need to test either of them (any is fine, as each of them carry the meaning of success VS failure), so you only have to choose whether your prefer to test card or initError.

If you don't, they you need to test both of them, like you did. But then you need to go a bit further in the defensiveness. Since we can't trust the caller, it could very well be e.g. that they have set s.card but they have forgotten to set s.initError. In which case, your test is going to be true, but you're going to pass a nil error to handleError. So we would need to check that s.initError is not nil, etc..

Anyway, long story short, since we (2 programmers at arribada) are the caller in that case, I would choose trust us, and I would change L32 to test only for s.initError.